### ISSUE TITLE
Next Curriculum editor bugs

### ISSUE THREAD

**Issue by RandellDawson** [freeCodeCamp/freeCodeCamp#39348]
I figured we need to track known bugs with the new editor.  A few are listed below:

1. When an editable region marker is placed at the very top of the challenge seed code, the widget (upper and lower jaws) does not appear. See [step 66 of the Basic HTML Cat Photo App](https://github.com/freeCodeCamp/freeCodeCamp/blob/next-curriculum/curriculum/challenges/english/01-responsive-web-design/basic-html-cat-photo-app/part-66.md) for an example.
![image](https://user-images.githubusercontent.com/5313213/89073744-e9f08880-d32f-11ea-8145-bc5ebea25f6a.png)

2. If editor is resized to a smaller width, the editor does not fill out the remain spacing to the right.
![image](https://user-images.githubusercontent.com/5313213/89073983-579cb480-d330-11ea-8f6c-5d928e5bc85f.png)

3. If there is not enough code below an editable region, the editable region (upper and lower jaw) does not shift up toward the top of the editor.  I believe it was discussed in a Front 3 meeting that there is a config setting that allows scrolling up when there are not extra lines at the bottom.
![image](https://user-images.githubusercontent.com/5313213/89074191-cd088500-d330-11ea-8c79-38f14a1ccd54.png)

**Comment by gikf** [comment]
Inconsistent and/or unwanted behavior:
* Write passing code.
* Check code with tests, regardless of way it's done (either button one of buttons, Ctrl + Enter shortcut), but don't submit it.
* Change code.
* Pressing button "Submit your code and go to next challenge (Ctrl + Enter)" will mark challenge as passed and go to next one, regardless if changed code would pass tests. Ctrl + Enter and other button will run tests again first.

**Comment by ojeytonwilliams** [comment]
@gikf I can confirm that that's definitely not what we intended to do.  The state should get reset as soon as the preview updates (i.e. after the user stops typing).  Thanks for the report!

**Comment by ShaunSHamilton** [comment]
With latest `main`:
```console
Error in function toStartOfLine in ./src/templates/Challenges/classic/Editor.js:123
Cannot read property 'setStartPosition' of null TypeError:
Cannot read property 'setStartPosition' of null at toStartOfLine (webpack-internal:///./src/templates/Challenges/classic/Editor.js:170:16) at preventOverlap
```

Not major, but could lead to other bugs.

Steps:
1) Have seed be (without space, incorrect editor is used):
````markdown
# --seed--

## --seed-contents--

```html

--fcc-editable-region--

--fcc-editable-region--

```
````
2) Get annoyed by the _look_ of the Editor with space
3) Remove space at opening `--fcc-editable-region--` marker:
````markdown
# --seed--

## --seed-contents--

```html
--fcc-editable-region--

--fcc-editable-region--

```
````
4) Get annoyed at Gatsby complaining
5) Complain to Oliver about Gatsby complaining

**Note**: Any change to the position of the `--fcc-editable-region--` while app is running will cause the above error.

_It is late...my humour is dry_

**Comment by ShaunSHamilton** [comment]
One more thing:

Whilst testing #41797, I decided to see what it would be like to copy-paste the code, as some Campers develop locally to use their own editor. Instead of copying, I cut the code, and Gatsby errored out.
- Go to a challenge with code (`index.html`)
- Copy all of the code
- Delete
- Gatsby complains

**Comment by Nirajn2311** [comment]
Maybe the issue description could hold the list of issues that are put here in the form of checkboxes so its easier for any of us to know which of these issues are solved or not. Here's another thought, maybe each of these issues could be put up as their own issue and a new label named "Next curriculum bugs" could be assigned to them. Basically what I'm saying is it feels cluttered to put all the issues in one place and it becomes harder to track or ask questions about the issues.

**Comment by ShaunSHamilton** [comment]
To add: The following code causes the following error
```css
h1, h2, p {text-align: center;}body {  background-color: brown;}
```

To reproduce:
1) Go to challenge with code inside and outside editable region
2) Delete lines from within/outside editable region, until code is on single line

![image](https://user-images.githubusercontent.com/51722130/120654809-f080ec80-c479-11eb-8062-63ab08e8bebc.png)

**Comment by Sembauke** [comment]
@ShaunSHamilton Yeah this is caused by the decorations going below the first line.

**Comment by ShaunSHamilton** [comment]
Briefly, on Chat, we discussed removing _normalize.css_ ([ref](https://github.com/freeCodeCamp/freeCodeCamp/pull/42067#discussion_r657535970)).

The pros to removing it:
- Camper's iframe will mimick real-world output
- We do not need to keep normalize in mind, when designing lessons
  - No need to create work arounds
- Reduced bundle size (_enticing Ahmad and Oliver_)

Cons:
- We use(d) normalize because silly browsers display things differently (different defaults for `margin`, `padding`, etc.)
- _Insert breaking change in production_ 🤷‍♂️ 

Essence: We cannot write lessons like the one referenced that say _"Now, remove all the margin, and look at the preview..."_ because normalize adds the margin back - we have to override it.

---

@ojeytonwilliams You had the most discussion to say about this, but I think we got side-tracked with other editor diffs ([ref](https://github.com/freeCodeCamp/freeCodeCamp/pull/42195)) .

**Comment by ojeytonwilliams** [comment]
My position is: the closer we make the preview pane to how raw html renders, the better.  However, we do need to check that this doesn't "break" any of the commonly used browsers.

This is easily reversed, though, so I'm fine with trying it (once we've done some QA on Chrome, Safari and Firefox)

**Comment by moT01** [comment]
Not sure if anyone noticed or is aware, but I think the typescript migration or something, maybe a package update, got rid of the jaws - I was reviewing the skyline project, they were there - I gave it a rebase, and they were gone. Checked out main and they aren't there either.

**Comment by Sembauke** [comment]
Should we revert the pr that converts the Editor into typescript, so we can see if the issue lays there?

**Comment by ojeytonwilliams** [comment]
@Sembauke no need, Tom's right that the migration caused the bug.  I've been looking into it and should have a fix before too long.

**Comment by ShaunSHamilton** [comment]
Probably related to: https://github.com/freeCodeCamp/freeCodeCamp/issues/39348#issuecomment-853880345

[Ref](https://github.com/freeCodeCamp/freeCodeCamp/pull/42855#issuecomment-879814529)
```
Uncaught Error: Illegal value for lineNumber
    at TextModel.getLineContent (textModel.js:521)
    at positionsToRange (editor.tsx:1021)
    at eval (editor.tsx:796)
    at Array.map (<anonymous>)
    at decorateForbiddenRanges (editor.tsx:795)
    at editorWillMount (editor.tsx:275)
    at MonacoEditor.editorWillMount (editor.js:118)
    at MonacoEditor.initMonaco (editor.js:109)
    at MonacoEditor.componentDidMount (editor.js:53)
```

Editable region defined at top of file causes it:
````markdown
## --seed-contents--

```html
--fcc-editable-region--
<html>
--fcc-editable-region--
  <head>
    <title>CatPhotoApp</title>
  </head>
````

Does not break:
````markdown
## --seed-contents--

```html

--fcc-editable-region--
<html>
--fcc-editable-region--
  <head>
    <title>CatPhotoApp</title>
  </head>
````

**Comment by Sembauke** [comment]
I think this is because of the top decoration displaying itself before the first line in the editor. When adding an extra line at the top this might not be the case. 

The decoration just seems to enter an "illegal line value"

**Comment by ojeytonwilliams** [comment]
That sounds very likely, Sem.  It's probably just trying to decorate line zero and crashing as a result.

**Comment by Sembauke** [comment]
Yes that's what I meant.

**Comment by gikf** [comment]
I'm looking at #43364 and getting this funny behavior.
![editor](https://user-images.githubusercontent.com/60067306/132131498-909a860b-77c7-417f-833b-35d7ad939f56.gif)

If initial editable region is empty:
- after entering first character cursor is moved to the left of it
- one line below gets pulled into the editable region
- order of the files at the top are swapped

If initial editable region is not empty only the last point occurs.

**Comment by gikf** [comment]
Different with jumpy cursor:
- Make both html and css visible
- After each character written in the left frame, cursor jumps to the right frame.
- It happens regardless where is the editable region.
![editor2](https://user-images.githubusercontent.com/60067306/136414777-436190a3-64f2-40e3-ac57-d41dc545fa9a.gif)

**Comment by ojeytonwilliams** [comment]
I think, tentatively, that these are all fixed now.

Once https://github.com/freeCodeCamp/freeCodeCamp/pull/43777 and https://github.com/freeCodeCamp/freeCodeCamp/pull/43728 I think this can be closed.  Then we can just open single issues as and when new bugs are discovered/created.
